Signal/src/network/GiphyDownloader.swift

```diff
@@ -0,0 +1,410 @@
+        self.rendition = rendition
+        self.filePath = filePath
+    }
+
+    deinit {
+        // Clean up on the asset on disk.
+        let filePathCopy = filePath
+        DispatchQueue.global().async {
+            do {
+                let fileManager = FileManager.default
+                try fileManager.removeItem(atPath:filePathCopy)
+            } catch let error as NSError {
+                owsFail("\(GiphyAsset.TAG) file cleanup failed: \(filePathCopy), \(error)")
+            }
+        }
+    }
+}
+
+// A simple LRU cache bounded by the number of entries.
+class LRUCache<KeyType: Hashable & Equatable, ValueType> {
+
+    private var cacheMap = [KeyType: ValueType]()
+    private var cacheOrder = [KeyType]()
+    private let maxSize: Int
+
+    init(maxSize: Int) {
+        self.maxSize = maxSize
+    }
+
+    public func get(key: KeyType) -> ValueType? {
+        guard let value = cacheMap[key] else {
+            return nil
+        }
+
+        // Update cache order.
+        cacheOrder = cacheOrder.filter { $0 != key }
+        cacheOrder.append(key)
+
+        return value
+    }
+
+    public func set(key: KeyType, value: ValueType) {
+        cacheMap[key] = value
+
+        // Update cache order.
+        cacheOrder = cacheOrder.filter { $0 != key }
+        cacheOrder.append(key)
+
+        while cacheOrder.count > maxSize {
+            guard let staleKey = cacheOrder.first else {
+                owsFail("Cache ordering unexpectedly empty")
+                return
+            }
+            cacheOrder.removeFirst()
+            cacheMap.removeValue(forKey:staleKey)
+        }
+    }
+}
+
+private var URLSessionTaskGiphyAssetRequest: UInt8 = 0
+
+// This extension is used to punch an asset request onto a download task.
+extension URLSessionTask {
+    var assetRequest: GiphyAssetRequest {
+        get {
+            return objc_getAssociatedObject(self, &URLSessionTaskGiphyAssetRequest) as! GiphyAssetRequest
+        }
+        set {
+            objc_setAssociatedObject(self, &URLSessionTaskGiphyAssetRequest, newValue, objc_AssociationPolicy.OBJC_ASSOCIATION_RETAIN_NONATOMIC)
+        }
+    }
+}
+
+@objc class GiphyDownloader: NSObject, URLSessionTaskDelegate, URLSessionDownloadDelegate {
+
+    // MARK: - Properties
+
+    static let TAG = "[GiphyDownloader]"
+
+    static let sharedInstance = GiphyDownloader()
+
+    // A private queue used for download task callbacks.
+    private let operationQueue = OperationQueue()
+
+    // Force usage as a singleton
+    override private init() {}
+
+    deinit {
+        NotificationCenter.default.removeObserver(self)
+    }
+
+    private let kGiphyBaseURL = "https://api.giphy.com/"
+
+    private func giphyDownloadSession() -> URLSession? {
+        // TODO: We need to verify that this session configuration properly
+        //       proxies all requests.
+        let configuration = URLSessionConfiguration.ephemeral
+        configuration.connectionProxyDictionary = [
```

Comment:
This looks like duplicated configuration - Can we extract the duplicated parts of this configuration with the giphy searcher?

Reply:
The configuration is similar but not quite the same.  We could constantize some of these values I suppose.